### DISCUSSION TITLE
Unit types

### DISCUSSION THREAD

**Discussion by MarcinZiabek** [QuestPDF/QuestPDF#98] [Ideas]
### Problem

In QuestPDF, there are multiple elements that expect a length value as an input, e.g. border (thickness), padding, Row (constant item width), etc. The library follows the default length unit from PDF specification: points. Where 72 points are equal to 1 inch. However, sometimes we need to define values in more physical units, e.g. inches, centimetres, etc. 

### Solutions

I think that QuestPDF can provide a simple API to simplify such workflow. I recognize three possibilities.

1) **Classic API with static methods.** This is similar to the API that is available for the TimeSpan struct, e.g. TimeSpan.FromSeconds(5).

```csharp
container
    .Width(MeasureUnit.Points(72))
    .Height(MeasureUnit.Inches(1))
```

2) **FluentAPI that extends the integer and float types.** Such concept was used in the FluentAssertions library to describe time periods. 

```csharp
container
    .Width(72.Points())
    .Height(1.Inches())
```

3) **Using arithmetic to combine values and units.** This is not very pragmatic, how many programmers can adapt to such an approach?

```csharp
// may look nice when change MeasureUnit into something shorter
container
    .Width(72 * MeasureUnit.Centimetre)
    .Height(3.5f * MeasureUnit.Inch)

// or even
container
    .Width(3.5f * MeasureUnit.Inch + 72 * MeasureUnit.Centimetre)
```

4) **Providing unit type as a second/optional argument.** This idea would be to extend certain methods of the fluent API implementation, with no breaking changes. This is very straightforward and discoverable.

```csharp
container
    .Width(72, Unit.Centimetre)
    .Height(3.5f, Unit.Inch)
```

### Constraints

The target structure can implement an implicit conversion operator. That means, when using only numbers, it will be interpreted as points. This solves the backwards compatibility problem as well as keeps the code clean. All lines in this example are equal:

```csharp
.Width(36.Points())
.Width(0.5f.Inches()) // looks strange?
.Width(36)
```

This API should be accepted only in applicable places. For example, there is no sense to use Inches when defining relative columns in Table element.

```csharp
table.ColumnsDefinition(columns =>
{
    columns.RelativeColumn(2.Inches()); // wrong!
    
    columns.ConstantColumn(75);
    columns.ConstantColumn(75);
    
    columns.ConstantColumn(75);
    columns.ConstantColumn(75);
});
```

Moreover, there are not many useful units in the document generation domain. I see the following:
- Point,
- Millimetre,
- Centimetre,
- Meter,
- Inch,
- Mils (1/1000 of an inch),
- Feet.

@marcianobandeira @qcz @lmingle @jcl-aadlab @donmurta @adaxer @pha3z @ebussa @teebeast @ahmedhsilverkey @eabtvc @wesleigomes I am allowing myself to mention you, hoping you will find some time to discuss those features 😁

**Comment by adaxer** [comment]
I think it's very useful. In a project that uses QuestPDF, I implemented a similar thing. But I stuck to the internal unit of points that QuestPDF (and pdf spec too) uses. So I introduced some extension functions that I call like eg. 

`page => page.Margin(10.MMToPoints())`

Like this, it would not become too complicated, the user can use whatever unit they want and use the suitable conversions like MMToPoints(), CMToPoints(), InToPoints() etc. And it fits well with the fluent style of QuestPDF.

**Reply by teebeast** [reply]
I would prefer `10.Mm().ToPoints()`. It's more flexible when converting between different units.

**Comment by qcz** [comment]
1 & 2 together would be very nice (the fluent API building on the classic).

The classic one is very similar to how one of the commercial reporting solution does it, so it would make porting very easy. The second one is very QuestPDF-y, so it would suit the library too.

**Comment by teebeast** [comment]
I would prefer a fluent api approach, like variant two.

``` csharp
page => page.Margin(10.Mm().ToPoints())
page => page.Margin(10.AsMm().ToPoints()) // Alternative prefixing with As when converting values to MessureUnit

```

Combining values where you can override arithmetic operations for MessureUnit:

``` csharp
page => page.Margin(10.Mm().ToPoints() + 7f.Inches())
```

**Comment by MarcinZiabek** [comment]
@adaxer @qcz @teebeast Thank you for your great responses!

I found one more possibility that is to provide unit type as an argument. I have no idea how I missed this 😂 Please take a look at proposition 4.

I am also considering if unit names should use abbreviations. `Millimetre` sounds like much but `Mm` is not very clear. I am not a huge fan of shortening names. What is your feeling?

**Reply by teebeast** [reply]
I agree with you. `Millimetre` is much more clear and I would also prefer it over `Mm`.

**Reply by teebeast** [reply]
#4 is also acceptable and useful.

**Comment by lmingle** [comment]
We could potentially look at what has been done in the UnitsNet project (https://github.com/angularsen/UnitsNet) for inspiration with conversion between units of measurement. The project defines LengthUnit with DtpPoint and has conversions between measurements units e.g. millimeters, inches along.

```
var inches = Length.FromInches(3.4);
var picas = inches.DtpPicas; // returns double 20.400000000004759
var points = inches.DtpPoints; // returns double 244.79999999999998

var millimeters = Length.FromMillimeters(65);
var picasMm = millimeters.DtpPicas; // returns double 15.354330708665
var pointsMm = millimeters.DtpPoints; // returns double 184.25196850393704
```

**Comment by wesleigomes** [comment]
I agree with most. The more explicit a command is, the better. I think it's
good practice what Laurence wrote:

var inches = Length.FromInches(3.4); var picas = inches.DtpPicas; //
returns double 20.400000000004759 var points = inches.DtpPoints; //(...)


Em seg., 17 de jan. de 2022 às 15:52, Laurence Mingle <
***@***.***> escreveu:

> We could potentially look at what has been done in the UnitsNet project (
> https://github.com/angularsen/UnitsNet) for inspiration with conversion
> between units of measurement. The project defines LengthUnit with DtpPoint
> and has conversions between measurements units e.g. millimeters, inches
> along.
>
> var inches = Length.FromInches(3.4);
> var picas = inches.DtpPicas; // returns double 20.400000000004759
> var points = inches.DtpPoints; // returns double 244.79999999999998
>
> var millimeters = Length.FromMillimeters(65);
> var picasMm = millimeters.DtpPicas; // returns double 15.354330708665
> var pointsMm = millimeters.DtpPoints; // returns double 184.25196850393704
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/QuestPDF/QuestPDF/discussions/98#discussioncomment-1985642>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD65AQAVYNLUFEPMM2QWDLDUWRQPDANCNFSM5MDDOCWQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Weslei Gomes
MSc, MBA, PMP

**Comment by MarcinZiabek** [comment]
> I agree with most. The more explicit a command is, the better. I think it's good practice what Laurence wrote

There is one thing that I don't like about solutions 1, 2 and 3. They all feel very complicated compared to the requirement. In fact, the only thing underneath is multiplication! Multiplying provided number by some constant.

My expectation is that once I call `Length.FromInches(3.4f)` of `5.Inches()` or `3.1f * Units.Inch`, I get an object that is applicable only in the length domain (e.g. thickness, padding, width), not just a number. This introduces a need to add a contract/interface to describe units. Something that idea 3 takes even higher by introducing basic units arithmetic. 

On the other hand, idea 4 seems simple and elegant. It also uses enums that can be statically imported if commonly used, giving us invocations such us `.Width(5, Inch)`. 

At the end of the day, we are aiming for a simple helper, aren't we? So developers don't need to pass around unit conversion constants, etc.

**Comment by ebussa** [comment]
1. `.Width(MeasureUnit.Points(72))` - clear, with static using could be shortened to `.Width(Points(72))` but is not "natural language"
2. `.Width(72.Points())` - not obvious, namespace pollution on int & float?, sometimes awkward: `0.5f.Inches()`
3. `.Width(35 * MeasureUnit.Centimetre)` - not obvious, confusing
4. `.Width(35, Unit.Centimetre)` - clear, could be shortened to `.Width(35, Centimetre)`

I think #1 & #4 are obvious, short and concise.  However #4 offers natural language `(72, Point)` instead `Point(72)`.  I think #2 & #3 are clever solutions, but both have a negative aspect and I'm not sure they offer any other advantage?

**Comment by girlpunk** [comment]
Just my 2c, but for many places in QuestPDF it's possible to follow the [typography pattern example](https://www.questpdf.com/documentation/api-reference.html#typography-pattern). For example, a `ReportConstants.TableSpacing` constant that sets the padding on various elements throughout a document. Such a constant wouldn't be possible with #4, but would with the other 3 solutions.

**Comment by MarcinZiabek** [comment]
> .Width(MeasureUnit.Points(72)) - clear, with static using could be shortened to .Width(Points(72)) but is not "natural language"
> .Width(72.Points()) - not obvious, namespace pollution on int & float?, sometimes awkward: 0.5f.Inches()
> .Width(35 * MeasureUnit.Centimetre) - not obvious, confusing
> .Width(35, Unit.Centimetre) - clear, could be shortened to .Width(35, Centimetre)

I also prefer option 4 for its simplicity. Implementation is not very difficult. However, it requires changing many methods (around 50 I guess) by adding a new optional argument that represents a unit type. 

> Just my 2c, but for many places in QuestPDF it's possible to follow the typography pattern example. For example, a ReportConstants.TableSpacing constant that sets the padding on various elements throughout a document. Such a constant wouldn't be possible with #4, but would with the other 3 solutions.

This is very true! If we follow option 4, it may be nice to cover option 1 as well. So developers can still easily convert different units into points.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
